File: tests/system/test_boot.py
"""Tests for kod/system/boot.py (Phase 2)."""

import pytest


class TestBootManagement:
    """Test boot operations."""

    def test_create_boot_entry_hook_callable(self):
        """create_boot_entry_hook() is callable."""
        from kod.system.boot import create_boot_entry_hook
        
        assert callable(create_boot_entry_hook)

    def test_get_kernel_version_callable(self):
        """get_kernel_version() is callable."""
        from kod.system.boot import get_kernel_version
        
        assert callable(get_kernel_version)

    def test_read_root_device(self, tmp_path):
        fstab = tmp_path / "fstab"
        fstab.write_text("# comment\nUUID=abc-123 / btrfs defaults 0 0\nUUID=def-456 /boot vfat 0 0\n")
        from kod.system.boot import _read_root_device
        assert _read_root_device(str(fstab)) == "UUID=abc-123"

    def test_create_boot_entry_hook_writes_files(self, tmp_path, monkeypatch):
        from kod.system import boot
        monkeypatch.setattr(boot, "get_kernel_file", lambda mp, package="linux": ("/x/vmlinuz", "6.12.0-arch1"))
        monkeypatch.setattr(boot, "exec", lambda cmd, get_output=False: "2026-01-01 00:00:00")
        (tmp_path / "etc").mkdir()
        (tmp_path / "etc/fstab").write_text("UUID=root-uuid / btrfs defaults 0 0\n")
        boot.create_boot_entry_hook(7, "linux", str(tmp_path))()
        entry = (tmp_path / "boot/loader/entries/kodos-7.conf").read_text()
        assert "root=UUID=root-uuid rw rootflags=subvol=generations/7/rootfs" in entry
        assert "linux /vmlinuz-6.12.0-arch1" in entry
        assert "initrd /initramfs-linux-6.12.0-arch1.img" in entry
        loader = (tmp_path / "boot/loader/loader.conf").read_text()
        assert "default kodos-7.conf" in loader

File: tests/system/test_boot.lua
-- Integration tests for kod.system.boot Lua module
-- Tests that Lua boot operations produce same output as Python versions

local boot = require('kod.system.boot')

local function test_read_root_device()
    -- Create temp fstab
    local fstab = "/tmp/test_fstab_" .. os.time()
    local f = io.open(fstab, "w")
    f:write("# comment\n")
    f:write("UUID=abc-123 / btrfs defaults 0 0\n")
    f:write("UUID=def-456 /boot vfat 0 0\n")
    f:close()
    
    local root = boot.read_root_device(fstab)
    assert(root == "UUID=abc-123", "Expected UUID=abc-123, got " .. root)
    
    os.remove(fstab)
    print("✓ test_read_root_device passed")
end

local function test_read_root_device_no_entry()
    local fstab = "/tmp/test_fstab_no_root_" .. os.time()
    local f = io.open(fstab, "w")
    f:write("UUID=def-456 /boot vfat 0 0\n")
    f:close()
    
    local ok, err = pcall(function()
        boot.read_root_device(fstab)
    end)
    assert(not ok, "Should have raised error for missing root")
    assert(err:match("No '/' entry found"), "Error message should mention missing root entry")
    
    os.remove(fstab)
    print("✓ test_read_root_device_no_entry passed")
end

local function test_read_root_device_with_extra_whitespace()
    -- Test that parser handles extra whitespace correctly
    local fstab = "/tmp/test_fstab_whitespace_" .. os.time()
    local f = io.open(fstab, "w")
    f:write("  UUID=root-uuid    /    btrfs   defaults   0   0  \n")
    f:write("UUID=boot-uuid /boot vfat 0 0\n")
    f:close()
    
    local root = boot.read_root_device(fstab)
    assert(root == "UUID=root-uuid", "Expected UUID=root-uuid with whitespace, got " .. root)
    
    os.remove(fstab)
    print("✓ test_read_root_device_with_extra_whitespace passed")
end

-- Run all tests
test_read_root_device()
test_read_root_device_no_entry()
test_read_root_device_with_extra_whitespace()

print("\nAll Lua boot tests passed!")
